### DISCUSSION TITLE
How to store information in a parse_tree::node

### DISCUSSION THREAD

**Discussion by mik30s** [taocpp/PEGTL#109] [Q&A] [answered]
Is it possible to store information in a `parse_tree::node`. I see there is a `has_content()` function but i'm not sure if i can store information in a node. For example during type checking in a compiler I would like to store a variable's type in its node for later reference. 

Thanks.

**Comment by d-frey** [comment, marked as answer]
Yes, that should be possible. There are two options: First you could use the provided node class and simply store a different type in it. We do this here:
https://github.com/taocpp/PEGTL/blob/f0b76a1b06f993306c5021c1e84b73ad2d1813e0/src/example/pegtl/abnf2pegtl.cpp#L277-L289
That is the simple, but also limited way.

If you are looking for a more advanced solution, I recommend you write your own node class. The design of the parse-tree supports this. You can derive from the existing node class, overwrite some methods and add member as needed. Again, the above `abnf2pegtl.cpp` contains an example:
https://github.com/taocpp/PEGTL/blob/f0b76a1b06f993306c5021c1e84b73ad2d1813e0/src/example/pegtl/abnf2pegtl.cpp#L244-L249
although the above does not add more members and just overwrite an existing method from the node class.

Let me know if any of those options works for you or if you need something else.

**Comment by mik30s** [comment]
Thanks, I'll try the first one out and see how far it takes me.
